pkg/api/customization/globaldns/action_globaldns.go

```diff
@@ -100,44 +97,84 @@ func (w *Wrapper) addProjects(gDNS *v3.GlobalDNS, request *types.APIContext, inp
 	if err := ma.CheckCallerAccessToTargets(request, inputProjects, client.ProjectType, &client.Project{}); err != nil {
 		return err
 	}
-	var projectsToAdd []string
-	for _, p := range inputProjects {
-		if !existingProjects[p] {
-			projectsToAdd = append(projectsToAdd, p)
-		}
+
+	split := strings.SplitN(request.ID, ":", 2)
+	if len(split) != 2 {
+		return fmt.Errorf("incorrect global DNS ID")
 	}
-	return w.updateGDNS(gDNS, projectsToAdd, request, "addedProjects")
-}
 
-func (w *Wrapper) removeProjects(gDNS *v3.GlobalDNS, request *types.APIContext, existingProjects map[string]bool, inputProjects []string) error {
-	toRemoveProjects := make(map[string]bool)
-	var finalProjects []string
-	for _, p := range inputProjects {
-		toRemoveProjects[p] = true
-	}
-	for _, p := range gDNS.Spec.ProjectNames {
-		if !toRemoveProjects[p] {
-			finalProjects = append(finalProjects, p)
+	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		existingProjects := make(map[string]bool)
+		gDNS, err := w.GlobalDNSes.GetNamespaced(split[0], split[1], v1.GetOptions{})
+		if err != nil {
+			return false, err
 		}
+		for _, p := range gDNS.Spec.ProjectNames {
+			existingProjects[p] = true
+		}
+		if gDNS.Spec.MultiClusterAppName != "" {
```

Comment:
can we error out before calculating existingProjects? so we don't calculate it always